reco_utils/recommender/ripplenet/data_loader.py

```diff
@@ -0,0 +1,82 @@
+# This code is modified from RippleNet
+# Online code of RippleNet: https://github.com/hwwang55/RippleNet
+
+import collections
+import os
+import numpy as np
+
+
+def load_kg(kg_final):
+    """Standarize indexes for items and entities
+
+    Args:
+        kg_final (pd.DataFrame): knowledge graph converted with columns head,
+         relation and tail, with internal entity IDs
+
+    Returns:
+        n_entity (int): number of entities in KG
+        n_relation (int): number of relations in KG
+        kg (dictionary): KG in dictionary shape
+    """
+    print('reading KG file ...')
+
+    n_entity = len(set(kg_final.iloc[:, 0]) | set(kg_final.iloc[:, 2]))
+    n_relation = len(set(kg_final.iloc[:, 1]))
+
+    kg = collections.defaultdict(list)
+    for index, row in kg_final.iterrows():
+        kg[row["head"]].append((row["tail"], row["relation"]))
+
+    return n_entity, n_relation, kg
+
+
+def get_ripple_set(kg, user_history_dict, n_hop, n_memory):
+    """Build Ripple Set, dictionary for the related entities in the KG
+     given the paths of users, number of hops and memory
+
+    Args:
+        kg (dictionary): KG in dictionary shape
+        user_history_dict (dictionary): train rating data with positive ratings
```

Comment:
Does it mean that in the training data, there is no negative samples, i.e., 0?

Reply:
I improved the description